Nondeterminism in TestTokenAliasing

This test fails when run repeatedly. We can confirm by running:
```go
go test ./... --run=TestTokenAliasing -count=500
```

Test failures look like this:
```
--- FAIL: TestTokenAliasing (0.00s)
    tokens_test.go:549: 
                Error Trace:    /Users/ianwahbe/go/src/github.com/pulumi/pulumi-terraform-bridge/pkg/tfbridge/tokens_test.go:549
                Error:          Not equal: 
                                expected: "{\n    \"auto-aliasing\": {\n        \"resources\": {\n            \"pkg_mod1_r1\": {\n                \"current\": \"pkg:mod1/r1:R1\",\n                \"past\": [\n                    {\n                        \"name\": \"pkg:index/mod1R1:Mod1R1\",\n                        \"inCodegen\": true,\n                        \"majorVersion\": 1\n                    }\n                ]\n            },\n            \"pkg_mod1_r2\": {\n                \"current\": \"pkg:mod1/r2:R2\",\n                \"past\": [\n                    {\n                        \"name\": \"pkg:index/mod1R2:Mod1R2\",\n                        \"inCodegen\": true,\n                        \"majorVersion\": 1\n                    }\n                ]\n            },\n            \"pkg_mod2_r1\": {\n                \"current\": \"pkg:mod2/r1:R1\",\n                \"past\": [\n                    {\n                        \"name\": \"pkg:index/mod2R1:Mod2R1\",\n                        \"inCodegen\": true,\n                        \"majorVersion\": 1\n                    }\n                ]\n            }\n        },\n        \"datasources\": {\n            \"pkg_mod1_r1\": {\n                \"current\": \"pkg:index/getMod1R1:getMod1R1\",\n                \"past\": [\n                    {\n                        \"name\": \"pkg:index/getMod1R1:getMod1R1\",\n                        \"inCodegen\": false,\n                        \"majorVersion\": 1\n                    }\n                ]\n            }\n        }\n    }\n}"
                                actual  : "{\n    \"auto-aliasing\": {\n        \"resources\": {\n            \"pkg_mod1_r1\": {\n                \"current\": \"pkg:mod1/r1:R1\",\n                \"past\": [\n                    {\n                        \"name\": \"pkg:index/mod1R1:Mod1R1\",\n                        \"inCodegen\": true,\n                        \"majorVersion\": 1\n                    }\n                ]\n            },\n            \"pkg_mod1_r2\": {\n                \"current\": \"pkg:mod1/r2:R2\",\n                \"past\": [\n                    {\n                        \"name\": \"pkg:index/mod1R2:Mod1R2\",\n                        \"inCodegen\": true,\n                        \"majorVersion\": 1\n                    }\n                ]\n            },\n            \"pkg_mod2_r1\": {\n                \"current\": \"pkg:mod2/r1:R1\",\n                \"past\": [\n                    {\n                        \"name\": \"pkg:index/mod2R1:Mod2R1\",\n                        \"inCodegen\": true,\n                        \"majorVersion\": 1\n                    }\n                ]\n            }\n        },\n        \"datasources\": {\n            \"pkg_mod1_r1\": {\n                \"current\": \"pkg:index/getMod1R1:getMod1R1\",\n                \"past\": [\n                    {\n                        \"name\": \"pkg:index/getMod1R1:getMod1R1\",\n                        \"inCodegen\": false,\n                        \"majorVersion\": 1\n                    }\n                ]\n            },\n            \"pkg_mod1_r1_legacy\": {\n                \"current\": \"pkg:index/getMod1R1:getMod1R1\",\n                \"majorVersion\": 1\n            }\n        }\n    }\n}"
                            
                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -44,2 +44,6 @@
                                                 ]
                                +            },
                                +            "pkg_mod1_r1_legacy": {
                                +                "current": "pkg:index/getMod1R1:getMod1R1",
                                +                "majorVersion": 1
                                             }
                Test:           TestTokenAliasing
                Messages:       No changes should imply no change in history
FAIL
FAIL    github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfbridge       0.897s
ok      github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfbridge/x     1.220s
ok      github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfgen  1.863s [no tests to run]
ok      github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfshim/sdk-v1  1.514s [no tests to run]
ok      github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfshim/sdk-v2  1.545s [no tests to run]
ok      github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfshim/tfplugin5       1.603s [no tests to run]
ok      github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfshim/walk    1.609s [no tests to run]
ok      github.com/pulumi/pulumi-terraform-bridge/v3/unstable/metadata  1.687s [no tests to run]
ok      github.com/pulumi/pulumi-terraform-bridge/v3/unstable/propertyvalue     1.672s [no tests to run]
FAIL
```

--- a/pkg/tfbridge/auto_aliasing.go
+++ b/pkg/tfbridge/auto_aliasing.go
@@ -155,20 +155,24 @@ func (info *ProviderInfo) ApplyAutoAliases() error {
 	// Applying resource aliases adds new resources to providerInfo.Resources. To keep
 	// this process deterministic, we don't apply resource aliases until all resources
 	// have been examined.
-	applyResourceAliases := []func(){}
+	//
+	// The same logic applies to datasources.
+	applyAliases := []func(){}
 
 	for tfToken, computed := range info.Resources {
 		r, _ := rMap.GetOk(tfToken)
-		aliasResource(info, r, &applyResourceAliases, hist.Resources,
+		aliasResource(info, r, &applyAliases, hist.Resources,
 			computed, tfToken, currentVersion)
-	}
-	for _, f := range applyResourceAliases {
-		f()
 	}
 
 	for tfToken, computed := range info.DataSources {
 		ds, _ := dMap.GetOk(tfToken)
-		aliasDataSource(info, ds, hist.DataSources, computed, tfToken, currentVersion)
+		aliasDataSource(info, ds, &applyAliases, hist.DataSources,
+			computed, tfToken, currentVersion)
+	}
+
+	for _, f := range applyAliases {
+		f()
 	}
 
 	if err := md.Set(artifact, aliasMetadataKey, hist); err != nil {
@@ -433,6 +437,7 @@ func aliasOrRenameResource(
 func aliasDataSource(
 	p *ProviderInfo,
 	ds shim.Resource,
+	queue *[]func(),
 	hist map[string]*tokenHistory[tokens.ModuleMember],
 	computed *DataSourceInfo,
 	tfToken string,
@@ -447,7 +452,8 @@ func aliasDataSource(
 			MajorVersion: version,
 		}
 	} else {
-		aliasOrRenameDataSource(p, computed, tfToken, prev, version)
+		*queue = append(*queue,
+			func() { aliasOrRenameDataSource(p, computed, tfToken, prev, version) })
 	}
 
 	if ds == nil {

--- a/pkg/tfbridge/tokens_test.go
+++ b/pkg/tfbridge/tokens_test.go
@@ -454,6 +454,16 @@ func makeAutoAliasing(t *testing.T) (
 }
 
 func TestTokenAliasing(t *testing.T) {
+	// We run this test multiple times to guard against nondeterminism.
+	//
+	// See https://github.com/pulumi/pulumi-terraform-bridge/issues/1286.
+	for i := 0; i < 500; i++ {
+		t.Run(fmt.Sprintf("%d", i), testTokenAliasing)
+	}
+}
+
+func testTokenAliasing(t *testing.T) {
+	t.Parallel()
 	provider := func() *tfbridge.ProviderInfo {
 		return &tfbridge.ProviderInfo{
 			P: (&schema.Provider{
@@ -463,11 +473,12 @@ func TestTokenAliasing(t *testing.T) {
 					"pkg_mod2_r1": nil,
 				},
 				DataSourcesMap: schema.ResourceMap{
-					"pkg_mod1_r1": nil,
+					"pkg_mod1_d1": nil,
 				},
 			}).Shim(),
 		}
 	}
+
 	simple := provider()
 
 	metadata, autoAliasing := makeAutoAliasing(t)
@@ -527,11 +538,11 @@ func TestTokenAliasing(t *testing.T) {
 		},
 	}, modules.Resources)
 	assert.Equal(t, map[string]*tfbridge.DataSourceInfo{
-		"pkg_mod1_r1": {Tok: "pkg:mod1/getR1:getR1"},
-		"pkg_mod1_r1_legacy": {
-			Tok:                "pkg:index/getMod1R1:getMod1R1",
-			DeprecationMessage: "pkg.index/getmod1r1.getMod1R1 has been deprecated in favor of pkg.mod1/getr1.getR1",
-			Docs:               &tfbridge.DocInfo{Source: "kg_mod1_r1.html.markdown"},
+		"pkg_mod1_d1": {Tok: "pkg:mod1/getD1:getD1"},
+		"pkg_mod1_d1_legacy": {
+			Tok:                "pkg:index/getMod1D1:getMod1D1",
+			DeprecationMessage: "pkg.index/getmod1d1.getMod1D1 has been deprecated in favor of pkg.mod1/getd1.getD1",
+			Docs:               &tfbridge.DocInfo{Source: "kg_mod1_d1.html.markdown"},
 		},
 	},
 		modules.DataSources)

